// @progress/kendo-react-layout/contextmenu/ContextMenu.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import * as n from "react";
import { useId as u, getActiveElement as l, classNames as a } from "@progress/kendo-react-common";
import { Popup as f } from "@progress/kendo-react-popup";
import { Menu as i } from "../menu/components/Menu.mjs";
const r = (e) => {
  const o = n.useRef(null), c = u(e.id);
  return n.useEffect(() => {
    e.show && setTimeout(() => {
      const m = o.current && o.current.element, t = m && m.querySelector(".k-menu-item"), s = l(t && t.ownerDocument);
      t && t !== s && t.focus();
    });
  }, [e.show]), /* @__PURE__ */ n.createElement(f, { show: e.show, offset: e.offset, popupClass: "k-menu-popup k-overflow-visible", animate: !1 }, /* @__PURE__ */ n.createElement(
    i,
    {
      id: c,
      vertical: !0,
      ...e,
      ref: o,
      className: a("k-context-menu", e.className),
      onClose: e.onClose,
      role: "menu"
    },
    e.children
  ));
};
r.displayName = "KendoReactContextMenu";
export {
  r as ContextMenu
};
