// @progress/kendo-react-inputs/colors/ColorPalette.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import * as i from "react";
import s from "prop-types";
import { ColorPaletteService as S } from "./utils/color-palette.service.mjs";
import { KEYS as l, validatePackage as w, classNames as u, kendoThemeMaps as E, getTabIndex as k, dispatchEvent as y, createPropsContext as R, withIdHOC as F, withPropsContext as N } from "@progress/kendo-react-common";
import { packageMetadata as P } from "../package-metadata.mjs";
import { PALETTEPRESETS as T } from "./models/palette-presets.mjs";
import { isPresent as I } from "./utils/misc.mjs";
import { parseColor as D } from "./utils/color-parser.mjs";
const z = 24, f = 10, L = "office", a = class a extends i.Component {
  constructor(t) {
    super(t), this.wrapperRef = i.createRef(), this.paletteService = null, this.focus = () => {
      this.wrapperRef && this.wrapperRef.current && this.wrapperRef.current.focus();
    }, this.onKeyDown = (e) => {
      switch (e.key) {
        case l.down:
          this.handleCellNavigation(0, 1);
          break;
        case l.up:
          this.handleCellNavigation(0, -1);
          break;
        case l.right:
          this.handleCellNavigation(1, 0);
          break;
        case l.left:
          this.handleCellNavigation(-1, 0);
          break;
        case l.enter:
          this.handleEnter(e);
          break;
        default:
          return;
      }
      e.preventDefault();
    }, this.onColorClick = (e, o) => {
      this.isUncontrolled ? this.setState({ selectedColor: e, focusedColor: e }) : this.setState({ focusedColor: e }), this.dispatchChangeEvent(e, o);
    }, this.onFocus = (e) => {
      this.paletteService && (this.setState({ focusedColor: this.state.selectedColor || this.paletteService.colorRows[0][0] }), this.props.onFocus && this.props.onFocus.call(void 0, e));
    }, this.onBlur = () => {
      this.setState({ focusedColor: void 0 });
    }, w(P), this.state = {
      selectedColor: this.props.value !== void 0 ? this.props.value : this.props.defaultValue ? this.props.defaultValue : void 0,
      isFirstRender: !0
    };
  }
  get guid() {
    return this.props.id;
  }
  /**
   * @hidden
   */
  render() {
    const t = this.getPaletteInfo(), e = this.paletteService = new S();
    e.setColorMatrix(t.colors, t.columns);
    const o = e.getCellCoordsFor(this.state.selectedColor), r = e.getCellCoordsFor(this.state.focusedColor);
    return t.colors.length ? /* @__PURE__ */ i.createElement(
      "div",
      {
        id: this.props.id,
        role: "grid",
        className: u(
          "k-colorpalette",
          {
            [`k-colorpalette-${E.sizeMap[this.props.size] || this.props.size}`]: this.props.size,
            "k-disabled": this.props.disabled
          },
          this.props.className
        ),
        onFocus: this.onFocus,
        onBlur: this.onBlur,
        onKeyDown: this.onKeyDown,
        "aria-disabled": this.props.ariaDisabled || (this.props.disabled ? "true" : void 0),
        "aria-labelledby": this.props.ariaLabelledBy,
        "aria-describedby": this.props.ariaDescribedBy,
        tabIndex: k(this.props.tabIndex, this.props.disabled),
        ref: this.wrapperRef
      },
      /* @__PURE__ */ i.createElement("table", { className: "k-colorpalette-table", role: "presentation" }, /* @__PURE__ */ i.createElement("tbody", { role: "rowgroup" }, this.renderRows(e.colorRows, o, r)))
    ) : "";
  }
  /**
   * @hidden
   */
  static getDerivedStateFromProps(t, e) {
    if (!e.isFirstRender && t.value !== void 0) {
      if (t.value === "" && e.selectedColor !== void 0)
        return { selectedColor: void 0 };
      if (t.value !== "" && t.value !== e.selectedColor)
        return { selectedColor: t.value };
    } else if (e.isFirstRender)
      return { isFirstRender: !1 };
    return null;
  }
  handleCellNavigation(t, e) {
    if (this.paletteService)
      if (this.focusedColorCooridanates) {
        const o = this.paletteService.getNextCell(
          this.focusedColorCooridanates,
          t,
          e
        );
        this.setState({ focusedColor: this.paletteService.getColorAt(o) });
      } else
        this.setState({ focusedColor: this.paletteService.colorRows[0][0] });
  }
  handleEnter(t) {
    this.isUncontrolled && this.setState({ selectedColor: this.state.focusedColor }), this.dispatchChangeEvent(this.state.focusedColor, t);
  }
  dispatchChangeEvent(t, e) {
    y(this.props.onChange, e, this, { value: t, rgbaValue: D(t, "rgba") });
  }
  get focusedColorCooridanates() {
    return this.state.focusedColor && this.paletteService ? this.paletteService.getCellCoordsFor(this.state.focusedColor) : void 0;
  }
  get isUncontrolled() {
    return this.props.value === void 0;
  }
  getPaletteInfo() {
    if (typeof this.props.palette == "string") {
      const t = T[this.props.palette];
      return I(t) ? {
        colors: t.colors,
        columns: this.props.columns || t.columns || f
      } : { colors: [], columns: 0 };
    }
    return { colors: this.props.palette || [], columns: this.props.columns || f };
  }
  renderRows(t, e, o) {
    return t.map((r, n) => /* @__PURE__ */ i.createElement("tr", { role: "row", key: n }, this.renderColumns(r, n, e, o)));
  }
  renderColumns(t, e, o, r) {
    const n = o !== void 0 && o.row === e, C = o && o.col, m = r !== void 0 && r.row === e, g = r && r.col;
    return t.map((d, c) => {
      const h = n && C === c, v = u("k-colorpalette-tile", {
        "k-selected": h,
        "k-focus": m && g === c
      });
      return /* @__PURE__ */ i.createElement(
        "td",
        {
          role: "gridcell",
          className: v,
          "aria-label": d,
          "aria-selected": h ? !0 : this.props.disabled ? void 0 : !1,
          style: { backgroundColor: d },
          onClick: (b) => this.onColorClick(d, b),
          id: this.createCellId({ row: e, col: c }),
          key: c
        }
      );
    });
  }
  createCellId(t) {
    return `${this.guid}_${t.row}_${t.col}`;
  }
};
a.displayName = "ColorPalette", a.propTypes = {
  palette: s.oneOfType([s.arrayOf(s.string.isRequired), s.string]),
  columns: s.number,
  tileSize: s.any,
  defaultValue: s.string,
  value: s.string,
  disabled: s.bool,
  tabIndex: s.number,
  onChange: s.func,
  onFocus: s.func,
  id: s.string,
  ariaLabelledBy: s.string,
  ariaDescribedBy: s.string,
  className: s.string,
  size: s.oneOf(["small", "medium", "large", null])
}, a.defaultProps = {
  palette: L,
  tileSize: z,
  size: "medium"
};
let p = a;
const U = R(), B = F(
  N(
    U,
    p
  )
);
B.displayName = "KendoReactColorPalette";
export {
  B as ColorPalette,
  U as ColorPalettePropsContext,
  p as ColorPaletteWithoutContext,
  f as DEFAULT_COLUMNS_COUNT,
  L as DEFAULT_PRESET,
  z as DEFAULT_TILE_SIZE
};
