// @progress/kendo-react-grid/GridClientContextReader.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import * as t from "react";
import { useId as f, useUnstyled as R } from "@progress/kendo-react-common";
import { useLocalization as p, useInternationalization as z } from "@progress/kendo-react-intl";
import { useGridState as C } from "./GridState.mjs";
import { cloneReactElement as I } from "./utils/index.mjs";
const U = (e) => {
  const o = f(), n = e.id || o, s = R(), c = p(), i = z(), [, r] = t.useReducer((m) => !m, !0), l = t.useRef([]), a = t.useRef(0), [u, , d] = C();
  return I(e.children, {
    ...u,
    ...d,
    id: n,
    unstyled: s,
    localization: c,
    intl: i,
    columnsRef: l,
    scrollLeftRef: a,
    forceUpdate: r,
    isClient: !0
  });
};
export {
  U as GridClientContextReader
};
