// @progress/kendo-react-charts/components/base/CollectionConfigurationComponent.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import * as s from "react";
import { ChartContext as h } from "../../ChartContext.mjs";
const p = class p extends s.Component {
  constructor(r, t) {
    super(r, t), this.optionsStore = t.optionsStore;
  }
  renderChildren(r, t) {
    const { children: o } = r.props, { _chartKey: c, _parentStore: e } = this.props, n = {
      ...r.props,
      _chartCollectionIdxKey: `${c}_${t}`,
      _parentStore: e
    };
    return s.cloneElement(r, n, o);
  }
  render() {
    const { _chartKey: r, _parentStore: t, children: o } = this.props;
    return (t || this.optionsStore).dispatch({
      chartKey: r,
      payload: []
    }), s.Children.map(o, (e, n) => s.isValidElement(e) ? this.renderChildren(e, n) : e);
  }
};
p.contextType = h;
let a = p;
export {
  a as CollectionConfigurationComponent
};
