// @progress/kendo-react-charts/tooltip/Point.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
class a {
  /**
   * @hidden
   */
  constructor(t, h) {
    this.value = t.value, this.category = t.category, this.categoryIndex = t.categoryIx, this.series = t.series, this.dataItem = t.dataItem, this.percentage = t.percentage, this.runningTotal = t.runningTotal, this.total = t.total, this.low = t.low, this.high = t.high, this.xLow = t.xLow, this.xHigh = t.xHigh, this.yLow = t.yLow, this.yHigh = t.yHigh, this.point = t, this.format = ((t.options || {}).tooltip || {}).format || h;
  }
  get formattedValue() {
    return this.format ? this.point.formatValue(this.format) : String(this.value);
  }
}
export {
  a as TooltipPoint
};
