// @progress/kendo-react-layout/bottomnavigation/BottomNavigation.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import * as o from "react";
import t from "prop-types";
import { validatePackage as T, focusFirstFocusableChild as $, useId as B, useDir as P, classNames as _, dispatchEvent as S, Keys as w } from "@progress/kendo-react-common";
import { packageMetadata as L } from "../package-metadata.mjs";
import { POSITION_MODE_CLASSES as A, ITEM_FLOW_CLASSES as H } from "./models/utils.mjs";
import { BottomNavigationItem as j } from "./BottomNavigationItem.mjs";
const I = o.forwardRef((a, p) => {
  T(L);
  const i = o.useRef(null), b = o.useCallback(() => {
    i.current && $(i.current);
  }, []), c = o.useCallback(
    () => ({
      element: i.current,
      focus: b
    }),
    [b]
  );
  o.useImperativeHandle(p, c);
  const {
    positionMode: v = s.positionMode,
    itemFlow: k = s.itemFlow,
    border: C = s.border,
    className: M,
    items: r,
    item: E,
    itemRender: N,
    disabled: m,
    style: O,
    id: y,
    onSelect: f,
    onKeyDown: h
  } = a, g = B(), F = P(i, a.dir), n = o.useMemo(() => a.fillMode === null || a.fill === null ? null : a.fill || a.fillMode || s.fillMode, [a.fillMode, a.fill]), u = o.useMemo(() => a.themeColor || s.themeColor, [a.themeColor]), D = o.useMemo(
    () => _(
      "k-bottom-nav",
      A[v],
      H[k],
      {
        [`k-bottom-nav-${n}`]: n,
        [`k-bottom-nav-${n}-${u}`]: !!(n && u),
        "k-bottom-nav-border": C,
        "k-disabled": m
      },
      M
    ),
    [v, u, n, k, C, m, M]
  ), d = o.useCallback(
    (e, l) => {
      r && !r[l].disabled && f && S(f, e, c(), {
        itemTarget: r[l],
        itemIndex: l
      });
    },
    [r, f]
  ), R = o.useCallback(
    (e, l) => {
      d(e, l);
    },
    [d]
  ), K = o.useCallback(
    (e, l) => {
      switch (e.keyCode) {
        case w.enter:
        case w.space:
          d(e, l), e.preventDefault();
          break;
      }
      S(
        h,
        e,
        c(),
        void 0
      );
    },
    [d, h]
  );
  return /* @__PURE__ */ o.createElement("nav", { ref: i, className: D, style: O, id: y || g, dir: F }, r && r.map((e, l) => /* @__PURE__ */ o.createElement(
    j,
    {
      ...e,
      key: l,
      index: l,
      id: `${y || g}-${l}`,
      disabled: m || e.disabled,
      selected: e.selected,
      dataItem: e,
      item: E,
      render: N,
      onSelect: R,
      onKeyDown: K
    }
  )));
}), s = {
  themeColor: "primary",
  fillMode: "flat",
  itemFlow: "vertical",
  positionMode: "fixed",
  border: !0
};
I.propTypes = {
  className: t.string,
  style: t.object,
  id: t.string,
  dir: t.string,
  themeColor: t.oneOf([
    "primary",
    "secondary",
    "tertiary",
    "info",
    "success",
    "warning",
    "error",
    "dark",
    "light",
    "inverse"
  ]),
  fill: t.oneOf(["solid", "flat"]),
  fillMode: t.oneOf(["solid", "flat"]),
  itemFlow: t.oneOf(["vertical", "horizontal"]),
  border: t.bool,
  disabled: t.bool,
  selected: t.number,
  onSelect: t.func
};
I.displayName = "KendoReactBottomNavigation";
export {
  I as BottomNavigation
};
