// @progress/kendo-react-buttons/ButtonGroup.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import * as e from "react";
import s from "prop-types";
import { validatePackage as k, useUnstyled as O, classNames as g, uButtonGroup as N } from "@progress/kendo-react-common";
import { packageMetadata as T } from "./package-metadata.mjs";
const h = (i) => {
  k(T);
  const { children: B, className: C, dir: c, disabled: p, width: n } = i, v = O(), d = i.unstyled || v, m = d && d.uButtonGroup, w = (t) => {
    const a = e.Children.count(t), l = c !== void 0 ? c === "rtl" : u.current && getComputedStyle(u.current).direction === "rtl" || !1;
    return e.Children.map(t, (o, r) => {
      if (e.isValidElement(o)) {
        const b = r === a - 1, f = l, R = g(
          o.props.className,
          N.position({
            c: m,
            start: f ? b : r === 0,
            end: f ? r === 0 : b
          })
        );
        return E(o, R);
      }
      return o;
    });
  }, E = (t, a) => {
    const l = { ...n ? { width: n } : {}, ...t.props.style || {} }, o = p || t.props.disabled, r = {
      ...t.props,
      ...a ? { className: a } : {},
      ...Object.keys(l).length ? { style: l } : {},
      ...o !== void 0 ? { disabled: o } : {}
    };
    return e.Children.count(t.props.children) > 0 ? e.cloneElement(t, r, t.props.children) : e.cloneElement(t, r);
  }, u = e.useRef(null), G = w(B), y = g(
    N.wrapper({
      c: m,
      stretched: !!n,
      disabled: p
    }),
    C
  ), P = {
    className: y,
    style: { width: `${n}` },
    dir: c,
    // Accessibility properties
    role: "group",
    "aria-disabled": p
  };
  return /* @__PURE__ */ e.createElement("div", { ref: u, ...P, className: y }, G);
};
h.propTypes = {
  children: s.oneOfType([s.arrayOf(s.element), s.element]),
  className: s.string,
  disabled: s.bool,
  width: s.string,
  dir: s.string
};
export {
  h as ButtonGroup
};
