// @progress/kendo-react-grid/header/client/GridFilterCellContainer.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import * as h from "react";
import { GridContext as u } from "../../GridClientWrapper.mjs";
import { cloneReactElement as c } from "../../utils/index.mjs";
const y = (t) => {
  const i = h.useContext(u), f = (e, r, o, C) => {
    if (!i.filterChange)
      return;
    const l = [];
    (e !== "" && e !== null || r !== "") && l.push({
      field: o,
      operator: r,
      value: e
    }), t.filter && t.filter.filters && (t.filter.filters || []).forEach((n) => {
      n && n.field !== o && l.push(n);
    });
    const a = t.filter && t.filter.logic ? t.filter.logic : "and";
    i.filterChange(l.length > 0 ? { logic: a, filters: l } : null, C);
  }, s = i.getCellPositionStyle(t.column), d = t.isCustom ? {
    style: { ...s }
  } : {};
  return i.isClient || t.isClient ? c(t.children, {
    render: t.render,
    onChange: (e) => {
      f(e.value, e.operator, t.field, e.syntheticEvent);
    },
    tdProps: { ...t.tdProps, ...d }
  }) : c(t.children, {});
};
export {
  y as GridFilterCellContainer
};
